### PR TITLE
Infinite scroll: Allow users to reach top/bottom of the log list before triggering scroll

### PR DESCRIPTION
**Pull request by matyax**
Addressing recent feedback, with these changes we allow users to reach the top or bottom of the log list, before an scrolling event triggers a new request.

The reason is that triggering a new requests when the users are in the process of scrolling is potentially confusing and disrupting. By debouncing and waiting a few milliseconds, we allow the user to stay in the desired state, and trigger a new request only if they scroll in the right direction.

Additional changes: code clean up for enhanced readability.

Part of https://github.com/grafana/grafana/issues/71728

**Special notes for your reviewer:**

Please give this a try and let me know how it feels. I recommend to simmulate different network conditions for an experience closer to the real world.

Demo:

https://github.com/grafana/grafana/assets/1069378/3d8bdcf0-6118-4ea0-87f1-c104734befd6

### CODE REVIEW

File: public/app/features/logs/components/InfiniteScroll.tsx [hunk 1]
```diff
@@ -57,8 +66,20 @@ export const InfiniteScroll = ({
         return;
       }
       event.stopImmediatePropagation();
+
+      // Give the user a chance to reach the top or bottom before triggering infinite scrolling requests
+      const onScrollingEdge = isOnEdge(lastScroll.current, scrollElement, onEdge, debouncedOnEdge, () => {
```
Comment (svennergr) [inline public/app/features/logs/components/InfiniteScroll.tsx:71]: Isn't the `onEdgeCallback` callback missing to call `setOnEdge(true)`?

**Review by svennergr — APPROVED** [review]
I really like that. I was wondering how it would feel if "scrolling away" would cancel the current request?
I.e. user scrolled up and triggered infinite scrolling by accident, scrolls down again to cancel it. That might be misleading for other use cases, but maybe worth an exploration?

File: public/app/features/logs/components/InfiniteScroll.tsx (hunk 1 above)
Comment (matyax) [reply public/app/features/logs/components/InfiniteScroll.tsx:71]: The debounced flavor always set it to true. There's no need to debounce setting to false, at least in this version.

Comment (svennergr) [reply public/app/features/logs/components/InfiniteScroll.tsx:71]: Oh right, I missed that!

**Comment by ivanahuckova** [comment]
I like the code cleanup and change, but quite often, when I scroll to the bottom, I don't see the loading message anymore:

https://github.com/grafana/grafana/assets/30407135/103ce49b-7ffb-40f1-bc7f-ac9da74c4cc2

**Comment by matyax** [comment]
>  quite often, when I scroll to the bottom, I don't see the loading message anymore:

That is true, great catch!

> I was wondering how it would feel if "scrolling away" would cancel the current request?

Good question. A new request should set `loading` in Explore to true, which disables scrolling behavior. If scrolling ever cancels a request, we can consider it a bug.

**Comment by matyax** [comment]
> but quite often, when I scroll to the bottom, I don't see the loading message anymore:

That should be fixed. @ivanahuckova if you can take a second look I would appreciate it a lot.

File: public/app/features/logs/components/InfiniteScroll.test.tsx [hunk 2]
```diff
@@ -73,63 +134,26 @@ describe('InfiniteScroll', () => {
 `);
   });
 
-  describe.each([LogsSortOrder.Descending, LogsSortOrder.Ascending])(
+  describe.each([LogsSortOrder.Descending])( //, LogsSortOrder.Ascending])(
```
Comment (ivanahuckova) [inline public/app/features/logs/components/InfiniteScroll.test.tsx:137]: Is this removed on purpose?

**Review by ivanahuckova — COMMENTED** [review]
The interaction is fixed and work as expected when it comes to triggering on bottom. 🙂

I find triggering of query when I reach the top not ideal as it moves the whole page and also triggers the request. Is it possible to just stop the outer scroll (page scroll that gets us to query editor)  if you are scrolling within the logs container? 

Overall, it improves the experience, so LGTM (once test question is resolved), but wondering if this is a final destination when it comes to top-scrolling.

File: public/app/features/logs/components/InfiniteScroll.test.tsx (hunk 2 above)
Comment (matyax) [reply public/app/features/logs/components/InfiniteScroll.test.tsx:137]: Leftover from testing!

**Comment by matyax** [comment]
Great feedback @ivanahuckova . I think we could sort it out with a live demo.

**Comment by matyax** [comment]
@ivanahuckova I totally see your point now. Would you mind giving it another try again? Also @svennergr 🙏 
Thanks!

Additionally, if you agree with the current behavior, feel free to give it the ✅ , as I would not have anything else to add to this PR.

**Comment by ivanahuckova** [comment]
@matyax it stopped working in some cases for me. See: 

https://github.com/grafana/grafana/assets/30407135/ab1e60cf-6a58-4c52-b359-03e838f5f9fa

**Comment by matyax** [comment]
> @matyax it stopped working in some cases for me. See:

That should not have changed. My last commit only affected scrolling to the top.

**Comment by svennergr** [comment]
> @matyax it stopped working in some cases for me. See:

I can confirm that. When scrolling away, and back to the top, infinite scrolling doesn't work intermittently.

Also, I wasn't sure if that was the case previously, but we now always scroll outside of the timerange. I guess that's foreseen for now, but did we decide how that behavior should look like? 
E.g. timerange is `now - now-5m` we can scroll into the past forever.

**Comment by matyax** [comment]
> E.g. timerange is now - now-5m we can scroll into the past forever

No, that should not ever happen, or only on a very specific edge case. I will double check that in https://github.com/grafana/grafana/pull/81725

Bottom line: we will probably not need this PR unless it's specifically asked, as it introduces unnecessary extra steps to trigger infinite scrolling. We can come back to this if required. Thanks all!

**Pull request closed without merging**